tech-specs/src/explore/design/ledger/pos-integration.md

```diff
@@ -0,0 +1,257 @@
+      - for each `slash in read(slash/{validator_address})`:
+        - if `bond_start <= slash.epoch && slash.epoch <= bond_end)`, `amount_after_slash *= (10_000 - slash.rate) / 10_000`
+      - credit the `amount_after_slash` to the `delegator_address` and debit the whole `amount` (before slash, if any) from the PoS account
+      - burn the slashed tokens (`amount - amount_after_slash`), if not zero
+
+### Other transactions
+
+- `submit_slashable_evidence(evidence)`:
+  - if `evidence in slash/{evidence.validator_address}`, panic
+  - validate the `evidence`
+  - append the `evidence` into `slash/{evidence.validator_address}`
+  - reduce the `validator/{validator_address}/total_deltas` for the `evidence.validator_address` by the slash rate in and before the `evidence.epoch`
+  - update the `validator/{validator_address}/voting_power` for the `evidence.validator_address` in and after epoch `n`
+  - update the `total_voting_power` in and after epoch `n`
+  - update `validator_set` in and after epoch `n`
+  - if there's a `parameters` change in any epoch after `n`, check that more than 2/3 of validators' voting power agree to the change or remove the queued up change if not
+
+## Validity predicate
+
+In the following description, "pre-state" is the state prior to transaction execution and "post-state" is the state posterior to it.
+
+Any changes to PoS epoched data are checked to update the structure as described in [epoched data storage](/explore/design/pos.md#storage).
+
+Because some key changes are expected to relate to others, the VP also accumulates some values that are checked for validity after key specific logic:
+- `validator_total_deltas: HashMap<Address, HashMap<Epoch, token::Change>>`
+- `validator_voting_powers: HashMap<Address, <HashMap<Epoch, u64>>`
+- `bond_deltas: HashMap<Address, HashMap<Epoch, Bond>>`
+- `unbond_deltas: HashMap<Address, HashMap<Epoch, Unbond>>`
+- `slashes: HashMap<Address, HashMap<Epoch, u8>>`
+- `validator_set_changes: HashSet<Epoch>`
+
+The accumulators are initialized to their default values (empty hash maps and hash set).
+
+All the above are keyed by validator addresses.
+
+The validity predicate triggers a validation logic based on the storage keys modified by a transaction.
+
+- `parameters`:
+  - if the `parameters` changed in epoch other than `n + unbonding_length`, panic
+  - check that more than 2/3 voting power agreed to the parameters change in `parameters/{parameters_hash}/{any_validator_address}` post-state
+- `parameters/{parameters_hash}`:
+  - check that the hash of value is equal to the `parameters_hash` in the key
+- `parameters/{parameters_hash}/{validator_address}`:
+  - check that the `parameters/{parameters_hash}` exists in the pre-state and post-state
+- `validator/{validator_address}/consensus_key`:
+  ```rust,ignore
+  match (pre_state, post_state) {
+    (None, Some(post)) => {
+      // - verify signature from tx.data against the post consensus key
```

Comment:
This should check the validator's validity predicate, which itself should check the signature,

Reply:
I removed this line for now in b72f7c92f03928e95a845342ae8a88cb8dad8458 and will update the validator's VP section with this instead